File: windows/API/StreamingAudioPlayer.py
"""
StreamingAudioPlayer - Real-time Audio Playback with Delay Support

This class provides high-performance audio playback using sounddevice with
support for delayed playback. It handles format conversion and provides
real-time buffer scheduling.

Key Features:
- Real-time audio buffer playback with minimal latency
- Configurable playback delay
- Automatic device selection
- Volume control and playback status monitoring
- Thread-safe buffer queue management
"""

import asyncio
import queue
import threading
import time
from datetime import datetime
from typing import Optional, Dict, Any, List
import numpy as np
import sounddevice as sd

from .AudioFormat import AudioFormat, AudioBuffer
from .AudioBufferQueue import AudioBufferQueue


class StreamingAudioPlayer:
    """Real-time audio player with streaming support"""
    
    def __init__(
        self,
        sample_rate: int = 48000,
        channels: int = 2,
        device_index: Optional[int] = None,
        delay: float = 0.0,
        blocksize: int = 1024,
        latency: str = 'low'
    ):
        """
        Initialize streaming audio player.
        
        Args:
            sample_rate: Sample rate in Hz
            channels: Number of channels
            device_index: sounddevice device index (None for default)
            delay: Playback delay in seconds
            blocksize: Audio block size
            latency: Latency setting ('low', 'high', or float value)
        """
        self.sample_rate = sample_rate
        self.channels = channels
        self.device_index = device_index
        self.delay = delay
        self.blocksize = blocksize
        self.latency = latency
        
        # State
        self._stream = None
        self._is_playing = False
        self._start_time = datetime.now()
        self._volume = 0.5
        
        # Delay management
        self._delay_start_time = None
        self._is_delay_active = delay > 0
        
        # Buffer queue for continuous playback
        self._buffer_queue = queue.Queue(maxsize=32)
        self._current_buffer = None
        self._current_position = 0
        
        # Statistics
        self._buffers_scheduled = 0
        self._buffers_played = 0
        self._underruns = 0
        
        # Debug logging
        self._debug_log_count = 0
        
        print(f"[{self._timestamp()}] StreamingAudioPlayer: Initialized")
        print(f"[{self._timestamp()}]   Sample Rate: {sample_rate}Hz")
        print(f"[{self._timestamp()}]   Channels: {channels}")
        print(f"[{self._timestamp()}]   Device: {self._get_device_name()}")
        print(f"[{self._timestamp()}]   Blocksize: {blocksize}")
        print(f"[{self._timestamp()}]   Latency: {latency}")
        if delay > 0:
            print(f"[{self._timestamp()}]   Playback delay: {delay}s")
    
    def _timestamp(self) -> str:
        """Get timestamp in milliseconds since player initialization"""
        elapsed = (datetime.now() - self._start_time).total_seconds() * 1000
        return f"{elapsed:07.1f}ms"
    
    def _get_device_name(self) -> str:
        """Get the name of the current output device"""
        if self.device_index is None:
            device_info = sd.query_devices(kind='output')
        else:
            device_info = sd.query_devices(self.device_index)
        return device_info['name'] if device_info else "Unknown"
    
    async def start_playback(self) -> None:
        """Start audio playback"""
        if self._is_playing:
            return
        
        # Don't start stream during delay period
        if self._is_delay_active and self._delay_start_time is None:
            self._delay_start_time = time.time()
            print(f"[{self._timestamp()}] StreamingAudioPlayer: Delay timer started - {self.delay}s")
            return
        
        # Create and start output stream
        self._stream = sd.OutputStream(
            samplerate=self.sample_rate,
            channels=self.channels,
            device=self.device_index,
            blocksize=self.blocksize,
            latency=self.latency,
            callback=self._audio_callback,
            finished_callback=self._finished_callback
        )
        
        self._stream.start()
        self._is_playing = True
        
        print(f"[{self._timestamp()}] StreamingAudioPlayer: Playback started")
    
    async def stop_playback(self) -> None:
        """Stop audio playback"""
        if not self._is_playing:
            return
        
        self._is_playing = False
        
        if self._stream:
            self._stream.stop()
            self._stream.close()
            self._stream = None
        
        # Clear buffer queue
        while not self._buffer_queue.empty():
            try:
                self._buffer_queue.get_nowait()
            except queue.Empty:
                break
        
        # Clear delay state
        self._delay_start_time = None
        self._is_delay_active = self.delay > 0
        
        print(f"[{self._timestamp()}] StreamingAudioPlayer: Playback stopped")
        print(f"[{self._timestamp()}]   Buffers played: {self._buffers_played}")
        print(f"[{self._timestamp()}]   Underruns: {self._underruns}")
    
    async def schedule_buffer(self, audio_data: np.ndarray) -> None:
        """
        Schedule an audio buffer for playback.
        
        Args:
            audio_data: Audio data as numpy array
        """
        # Handle delay
        if self._is_delay_active:
            if self._delay_start_time is None:
                self._delay_start_time = time.time()
                print(f"[{self._timestamp()}] StreamingAudioPlayer: Delay timer started - {self.delay}s")
                return
            
            elapsed = time.time() - self._delay_start_time
            if elapsed < self.delay:
                remaining = self.delay - elapsed
                if int(elapsed * 2) % 1 == 0:  # Log every 0.5s
                    print(f"[{self._timestamp()}] StreamingAudioPlayer: Delay in progress ({remaining:.1f}s remaining)")
                return
            else:
                # Delay expired - start playback
                if not self._is_playing:
                    print(f"[{self._timestamp()}] StreamingAudioPlayer: Delay expired - starting playback")
                    self._is_delay_active = False
                    await self.start_playback()
        
        # Ensure playback is started
        if not self._is_playing:
            await self.start_playback()
        
        # Convert format if needed
        if audio_data.dtype != np.float32:
            if audio_data.dtype == np.int16:
                audio_data = audio_data.astype(np.float32) / 32768.0
            else:
                audio_data = audio_data.astype(np.float32)
        
        # Apply volume
        audio_data = audio_data * self._volume
        
        # Ensure correct shape
        if audio_data.ndim == 1 and self.channels > 1:
            # Mono to multi-channel
            audio_data = np.tile(audio_data[:, np.newaxis], (1, self.channels))
        elif audio_data.ndim == 2 and audio_data.shape[1] != self.channels:
            # Channel count mismatch
            if audio_data.shape[1] > self.channels:
                audio_data = audio_data[:, :self.channels]
            else:
                # Duplicate last channel
                padding = self.channels - audio_data.shape[1]
                last_channel = audio_data[:, -1:]
                audio_data = np.hstack([audio_data] + [last_channel] * padding)
        
        # Debug logging for first few buffers
        if self._debug_log_count < 3:
            print(f"[{self._timestamp()}] StreamingAudioPlayer: Scheduling buffer #{self._debug_log_count + 1}")
            print(f"[{self._timestamp()}]   Shape: {audio_data.shape}")
            print(f"[{self._timestamp()}]   dtype: {audio_data.dtype}")
            self._debug_log_count += 1
        
        # Add to queue
        try:
            self._buffer_queue.put_nowait(audio_data)
            self._buffers_scheduled += 1
            
            if self._buffers_scheduled % 50 == 0:
                latency = self._buffers_scheduled - self._buffers_played
                print(f"[{self._timestamp()}] StreamingAudioPlayer: Scheduled {self._buffers_scheduled} buffers, latency: {latency}")
        
        except queue.Full:
            print(f"[{self._timestamp()}] StreamingAudioPlayer: Buffer queue full, dropping buffer")
    
    def _audio_callback(self, outdata, frames, time_info, status):
        """Audio stream callback"""
        if status:
            print(f"[{self._timestamp()}] StreamingAudioPlayer: Stream status: {status}")
            if status.output_underflow:
                self._underruns += 1
        
        # Fill output buffer
        samples_needed = frames
        write_position = 0
        
        while samples_needed > 0:
            # Get current buffer if needed
            if self._current_buffer is None or self._current_position >= len(self._current_buffer):
                try:
                    self._current_buffer = self._buffer_queue.get_nowait()
                    self._current_position = 0
                    self._buffers_played += 1
                except queue.Empty:
                    # No data available - fill with silence
                    outdata[write_position:] = 0
                    return
            
            # Copy data from current buffer
            available = len(self._current_buffer) - self._current_position
            to_copy = min(samples_needed, available)
            
            end_pos = self._current_position + to_copy
            outdata[write_position:write_position + to_copy] = \
                self._current_buffer[self._current_position:end_pos]
            
            self._current_position = end_pos
            write_position += to_copy
            samples_needed -= to_copy
    
    def _finished_callback(self):
        """Called when stream finishes"""
        print(f"[{self._timestamp()}] StreamingAudioPlayer: Stream finished")
    
    def set_volume(self, volume: float) -> None:
        """
        Set playback volume.
        
        Args:
            volume: Volume level (0.0 to 1.0)
        """
        self._volume = max(0.0, min(1.0, volume))
    
    def get_volume(self) -> float:
        """Get current volume"""
        return self._volume
    
    def set_delay(self, delay: float) -> None:
        """
        Set playback delay.
        
        Args:
            delay: Delay in seconds
        """
        self.delay = max(0.0, delay)
        self._is_delay_active = self.delay > 0
        
        if self._is_delay_active:
            print(f"[{self._timestamp()}] StreamingAudioPlayer: Playback delay set to {self.delay}s")
        else:
            print(f"[{self._timestamp()}] StreamingAudioPlayer: No playback delay")
    
    @property
    def is_active(self) -> bool:
        """Check if playback is active"""
        return self._is_playing
    
    def get_status(self) -> Dict[str, Any]:
        """Get playback status"""
        queue_size = self._buffer_queue.qsize()
        
        remaining_delay = 0.0
        if self._is_delay_active and self._delay_start_time:
            elapsed = time.time() - self._delay_start_time
            remaining_delay = max(0.0, self.delay - elapsed)
        
        return {
            "is_playing": self._is_playing,
            "volume": self._volume,
            "buffers_scheduled": self._buffers_scheduled,
            "buffers_played": self._buffers_played,
            "queued_buffers": queue_size,
            "underruns": self._underruns,
            "sample_rate": self.sample_rate,
            "channels": self.channels,
            "device": self._get_device_name(),
            "playback_delay": self.delay,
            "is_delay_active": self._is_delay_active,
            "remaining_delay": remaining_delay
        }
    
    async def play_test_tone(self, duration: float = 1.0, frequency: float = 440.0) -> None:
        """
        Play a test tone.
        
        Args:
            duration: Duration in seconds
            frequency: Tone frequency in Hz
        """
        print(f"[{self._timestamp()}] StreamingAudioPlayer: Generating {duration}s test tone at {frequency}Hz")
        
        # Generate sine wave
        t = np.linspace(0, duration, int(self.sample_rate * duration))
        tone = np.sin(2 * np.pi * frequency * t) * 0.3
        
        # Make stereo if needed
        if self.channels > 1:
            tone = np.tile(tone[:, np.newaxis], (1, self.channels))
        
        # Schedule in chunks
        chunk_size = self.blocksize * 4  # 4 blocks at a time
        for i in range(0, len(tone), chunk_size):
            chunk = tone[i:i + chunk_size]
            await self.schedule_buffer(chunk)
        
        # Wait for playback to complete
        await asyncio.sleep(duration + 0.5)
        await self.stop_playback()


class AudioStreamDelegate:
    """Protocol for audio stream delegates (for compatibility with macOS API)"""
    
    async def audio_streamer_did_receive(self, streamer, buffer: AudioBuffer):
        """Called when audio buffer is received"""
        pass
    
    async def audio_streamer_did_encounter_error(self, streamer, error: Exception):
        """Called when error occurs"""
        pass
    
    async def audio_streamer_did_finish(self, streamer):
        """Called when streaming finishes"""
        pass

File: macos/API/Sources/AudioCaptureKit/StreamingAudioPlayer.swift
import Foundation
import AVFoundation

/// StreamingAudioPlayer - Real-time Audio Playback with Delay Support
///
/// This class provides high-performance audio playback using AVAudioEngine with
/// support for delayed playback. It automatically routes audio to USB SPDIF adapter
/// and handles format conversion as needed.
///
/// Key Features:
/// - Real-time PCM buffer playback with minimal latency
/// - Configurable playback delay (audio engine starts after delay)
/// - Automatic USB SPDIF adapter detection and routing
/// - Direct AVAudioPCMBuffer scheduling for zero-copy performance
/// - Volume control and playback status monitoring
///
/// Usage:
/// ```swift
/// let player = StreamingAudioPlayer(delay: 2.0) // 2 second delay
/// try player.startPlayback()
/// player.scheduleBuffer(audioBuffer)
/// ```
///
/// Technical Details:
/// - Uses AVAudioPlayerNode for hardware-accelerated playback
/// - Implements AudioStreamDelegate to receive buffers from recorder
/// - Handles Float32 deinterleaved format natively
/// - Delay is implemented by postponing audio engine start

// MARK: - Streaming Audio Player (Direct AVAudioPCMBuffer Support)
@available(macOS 13.0, *)
public class StreamingAudioPlayer: NSObject {
    
    // MARK: - Properties
    
    private let audioEngine = AVAudioEngine()
    private let playerNode = AVAudioPlayerNode()
    private let mixerNode = AVAudioMixerNode()
    
    // Start time for precise timing
    private let startTime = Date()
    
    private var isEngineRunning = false
    private var isPlayerPlaying = false
    private var outputFormat: AVAudioFormat?
    
    // Buffer management for continuous playback
    private var bufferQueue: [AVAudioPCMBuffer] = []
    private let bufferQueueLock = NSLock()
    private let maxQueueSize = 8
    private var isSchedulingBuffers = false
    
    // Performance monitoring
    private var buffersScheduled = 0
    private var buffersPlayed = 0
    private var debugLogCount = 0
    
    // Delay functionality
    private var playbackDelay: TimeInterval = 0.0
    private var delayStartTime: Date?
    private var delayBufferQueue: [AVAudioPCMBuffer] = []
    private let delayBufferLock = NSLock()
    private var isDelayActive = false
    
    // MARK: - Utilities
    
    /// Get timestamp in milliseconds since player initialization
    private func timestamp() -> String {
        let elapsed = Date().timeIntervalSince(startTime) * 1000
        return String(format: "[%07.1fms]", elapsed)
    }
    
    // MARK: - Initialization
    
    override init() {
        super.init()
        setupAudioEngine()
    }
    
    convenience init(delay: TimeInterval) {
        self.init()
        setPlaybackDelay(delay)
    }
    
    deinit {
        stopPlayback()
    }
    
    // MARK: - Audio Engine Setup
    
    private func setupAudioEngine() {
        // Set the output device to USB SPDIF Adapter
        setOutputDevice()
        
        // Attach nodes to engine
        audioEngine.attach(playerNode)
        audioEngine.attach(mixerNode)
        
        // Get the engine's output format
        outputFormat = audioEngine.outputNode.outputFormat(forBus: 0)
        
        print("\(timestamp()) StreamingAudioPlayer: Engine output format: \(outputFormat!)")
        print("\(timestamp())   Sample Rate: \(outputFormat!.sampleRate)Hz")
        print("\(timestamp())   Channels: \(outputFormat!.channelCount)")
        print("\(timestamp())   Format: \(outputFormat!.commonFormat.rawValue)")
        print("\(timestamp())   Interleaved: \(outputFormat!.isInterleaved)")
        
        // Connect player to mixer, then to output
        audioEngine.connect(playerNode, to: mixerNode, format: outputFormat)
        audioEngine.connect(mixerNode, to: audioEngine.outputNode, format: nil)
        
        // Set initial volume
        playerNode.volume = 0.5
        
        print("\(timestamp()) StreamingAudioPlayer: Audio engine configured for direct PCM buffer streaming")
    }
    
    private func setOutputDevice() {
        // Find USB SPDIF Adapter device ID
        guard let usbSpdifDeviceID = findAudioDevice(named: "USB SPDIF Adapter") else {
            print("\(timestamp()) StreamingAudioPlayer: USB SPDIF Adapter not found, using default output")
            return
        }
        
        print("\(timestamp()) StreamingAudioPlayer: Setting output device to USB SPDIF Adapter (ID: \(usbSpdifDeviceID))")
        
        // Set the default output device
        var deviceID = usbSpdifDeviceID
        var propertyAddress = AudioObjectPropertyAddress(
            mSelector: kAudioHardwarePropertyDefaultOutputDevice,
            mScope: kAudioObjectPropertyScopeGlobal,
            mElement: kAudioObjectPropertyElementMain
        )
        
        let result = AudioObjectSetPropertyData(
            AudioObjectID(kAudioObjectSystemObject),
            &propertyAddress,
            0,
            nil,
            UInt32(MemoryLayout<AudioDeviceID>.size),
            &deviceID
        )
        
        if result == noErr {
            print("\(timestamp()) StreamingAudioPlayer: Successfully set USB SPDIF Adapter as output device")
        } else {
            print("\(timestamp()) StreamingAudioPlayer: Failed to set USB SPDIF Adapter as output device (error: \(result))")
        }
    }
    
    private func findAudioDevice(named targetName: String) -> AudioDeviceID? {
        var propertySize: UInt32 = 0
        var propertyAddress = AudioObjectPropertyAddress(
            mSelector: kAudioHardwarePropertyDevices,
            mScope: kAudioObjectPropertyScopeGlobal,
            mElement: kAudioObjectPropertyElementMain
        )
        
        AudioObjectGetPropertyDataSize(AudioObjectID(kAudioObjectSystemObject), &propertyAddress, 0, nil, &propertySize)
        let deviceCount = Int(propertySize) / MemoryLayout<AudioDeviceID>.size
        var devices = [AudioDeviceID](repeating: 0, count: deviceCount)
        
        AudioObjectGetPropertyData(AudioObjectID(kAudioObjectSystemObject), &propertyAddress, 0, nil, &propertySize, &devices)
        
        for device in devices {
            var name: CFString = "" as CFString
            var nameSize = UInt32(MemoryLayout<CFString>.size)
            var nameAddress = AudioObjectPropertyAddress(
                mSelector: kAudioDevicePropertyDeviceNameCFString,
                mScope: kAudioObjectPropertyScopeGlobal,
                mElement: kAudioObjectPropertyElementMain
            )
            
            let result = AudioObjectGetPropertyData(device, &nameAddress, 0, nil, &nameSize, &name)
            if result == noErr {
                let deviceName = name as String
                if deviceName == targetName {
                    return device
                }
            }
        }
        
        return nil
    }
    
    // MARK: - Playback Control
    
    public func startPlayback() throws {
        guard !isEngineRunning else { return }
        
        // Prepare and start the engine
        audioEngine.prepare()
        try audioEngine.start()
        isEngineRunning = true
        
        print("\(timestamp()) StreamingAudioPlayer: Audio engine started")
        
        // Start the player node
        if !playerNode.isPlaying {
            playerNode.play()
            isPlayerPlaying = true
            print("\(timestamp()) StreamingAudioPlayer: Player node started")
        }
    }
    
    public func stopPlayback() {
        if playerNode.isPlaying {
            playerNode.stop()
            isPlayerPlaying = false
        }
        
        if isEngineRunning {
            audioEngine.stop()
            isEngineRunning = false
        }
        
        // Clear buffer queue
        bufferQueueLock.lock()
        bufferQueue.removeAll()
        bufferQueueLock.unlock()
        
        // Clear delay state
        clearDelayState()
        
        print("\(timestamp()) StreamingAudioPlayer: Playback stopped")
    }
    
    // MARK: - Delay Management
    
    /// Set the playback delay in seconds
    /// - Parameter delay: Delay in seconds before audio playback starts
    func setPlaybackDelay(_ delay: TimeInterval) {
        delayBufferLock.lock()
        playbackDelay = max(0.0, delay)
        isDelayActive = playbackDelay > 0.0
        
        if isDelayActive {
            print("\(timestamp()) StreamingAudioPlayer: Playback delay set to \(playbackDelay) seconds")
        } else {
            print("\(timestamp()) StreamingAudioPlayer: No playback delay")
        }
        delayBufferLock.unlock()
    }
    
    /// Get the current playback delay
    var currentDelay: TimeInterval {
        return playbackDelay
    }
    
    /// Check if delay period has expired
    private func isDelayExpired() -> Bool {
        guard isDelayActive, let startTime = delayStartTime else { return true }
        return Date().timeIntervalSince(startTime) >= playbackDelay
    }
    
    /// Clean up delay state (no longer buffering - just timing delay)
    private func clearDelayState() {
        delayBufferLock.lock()
        delayBufferQueue.removeAll()
        delayStartTime = nil
        isDelayActive = false
        delayBufferLock.unlock()
    }
    
    // MARK: - Direct PCM Buffer Scheduling
    
    /// Schedule an audio buffer for playback
    /// During delay period, this method returns immediately without playing
    /// - Parameter buffer: The PCM audio buffer to play
    public func scheduleBuffer(_ buffer: AVAudioPCMBuffer) {
        guard isEngineRunning && isPlayerPlaying else {
            print("\(timestamp()) StreamingAudioPlayer: Cannot schedule buffer - engine not running")
            return
        }
        
        // Log buffer details (first few times)
        if debugLogCount < 3 {
            print("\(timestamp()) StreamingAudioPlayer: Scheduling PCM buffer #\(debugLogCount + 1)")
            print("\(timestamp())   Format: \(buffer.format)")
            print("\(timestamp())   Frame length: \(buffer.frameLength)")
            print("\(timestamp())   Frame capacity: \(buffer.frameCapacity)")
            debugLogCount += 1
        }
        
        // Check if format conversion is needed
        if let outputFormat = outputFormat, !buffer.format.isEqual(outputFormat) {
            // Convert buffer to output format
            if let convertedBuffer = convertBuffer(buffer, to: outputFormat) {
                scheduleConvertedBuffer(convertedBuffer)
            } else {
                print("\(timestamp()) StreamingAudioPlayer: Failed to convert buffer format")
            }
        } else {
            // Schedule buffer directly (optimal path)
            scheduleConvertedBuffer(buffer)
        }
    }
    
    private func scheduleConvertedBuffer(_ buffer: AVAudioPCMBuffer) {
        buffersScheduled += 1
        
        // Schedule the buffer for immediate playback
        playerNode.scheduleBuffer(buffer) { [weak self] in
            self?.bufferQueueLock.lock()
            self?.buffersPlayed += 1
            
            // Log performance metrics occasionally
            if let self = self, self.buffersPlayed % 50 == 0 {
                let latency = self.buffersScheduled - self.buffersPlayed
                print("\(self.timestamp()) StreamingAudioPlayer: Played \(self.buffersPlayed) buffers, queue latency: \(latency)")
            }
            
            self?.bufferQueueLock.unlock()
            
            // Schedule next buffer if available
            self?.processBufferQueue()
        }
    }
    
    private func processBufferQueue() {
        bufferQueueLock.lock()
        guard !bufferQueue.isEmpty else {
            bufferQueueLock.unlock()
            return
        }
        
        let nextBuffer = bufferQueue.removeFirst()
        bufferQueueLock.unlock()
        
        // Schedule next buffer
        scheduleConvertedBuffer(nextBuffer)
    }
    
    // MARK: - Format Conversion (If Needed)
    
    private func convertBuffer(_ inputBuffer: AVAudioPCMBuffer, to outputFormat: AVAudioFormat) -> AVAudioPCMBuffer? {
        // Create converter
        guard let converter = AVAudioConverter(from: inputBuffer.format, to: outputFormat) else {
            print("StreamingAudioPlayer: Failed to create audio converter")
            return nil
        }
        
        // Create output buffer
        guard let outputBuffer = AVAudioPCMBuffer(pcmFormat: outputFormat, frameCapacity: inputBuffer.frameLength) else {
            print("StreamingAudioPlayer: Failed to create output buffer")
            return nil
        }
        
        // Convert audio
        var error: NSError?
        let inputBlock: AVAudioConverterInputBlock = { inNumPackets, outStatus in
            outStatus.pointee = .haveData
            return inputBuffer
        }
        
        let status = converter.convert(to: outputBuffer, error: &error, withInputFrom: inputBlock)
        
        if let error = error {
            print("StreamingAudioPlayer: Conversion error: \(error)")
            return nil
        }
        
        if status == .haveData && outputBuffer.frameLength > 0 {
            return outputBuffer
        } else {
            print("StreamingAudioPlayer: Conversion failed - status: \(status)")
            return nil
        }
    }
    
    // MARK: - Buffer Queue Management
    
    private func addToBufferQueue(_ buffer: AVAudioPCMBuffer) {
        bufferQueueLock.lock()
        
        // Prevent buffer queue from growing too large (memory management)
        if bufferQueue.count >= maxQueueSize {
            // Remove oldest buffer
            bufferQueue.removeFirst()
            print("StreamingAudioPlayer: Buffer queue full, dropping oldest buffer")
        }
        
        bufferQueue.append(buffer)
        bufferQueueLock.unlock()
    }
    
    // MARK: - Volume Control
    
    var volume: Float {
        get { playerNode.volume }
        set { playerNode.volume = max(0.0, min(1.0, newValue)) }
    }
    
    // MARK: - Status
    
    var isPlaybackActive: Bool {
        return isEngineRunning && isPlayerPlaying
    }
    
    // MARK: - Test Functions
    
    func playTestTone(duration: Double) throws {
        try startPlayback()
        
        guard let format = outputFormat else {
            throw NSError(domain: "StreamingAudioPlayer", code: 1, 
                         userInfo: [NSLocalizedDescriptionKey: "No audio format available"])
        }
        
        let sampleRate = format.sampleRate
        let frequency: Float = 440.0 // A4 note
        let frameCount: AVAudioFrameCount = 1920 // 40ms at 48kHz
        
        print("StreamingAudioPlayer: Generating \(duration)s test tone at \(frequency)Hz")
        
        // Generate and schedule test tone buffers
        let totalFrames = AVAudioFrameCount(duration * sampleRate)
        var framesScheduled: AVAudioFrameCount = 0
        
        while framesScheduled < totalFrames {
            let framesToSchedule = min(frameCount, totalFrames - framesScheduled)
            
            guard let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: framesToSchedule) else {
                throw NSError(domain: "StreamingAudioPlayer", code: 2, 
                             userInfo: [NSLocalizedDescriptionKey: "Failed to create buffer"])
            }
            
            buffer.frameLength = framesToSchedule
            
            // Generate sine wave
            for channel in 0..<Int(format.channelCount) {
                guard let channelData = buffer.floatChannelData?[channel] else { continue }
                for frame in 0..<Int(framesToSchedule) {
                    let sampleIndex = framesScheduled + AVAudioFrameCount(frame)
                    let t = Float(sampleIndex) / Float(sampleRate)
                    channelData[frame] = sin(2.0 * .pi * frequency * t) * 0.3
                }
            }
            
            scheduleBuffer(buffer)
            framesScheduled += framesToSchedule
        }
        
        // Wait for completion
        let semaphore = DispatchSemaphore(value: 0)
        DispatchQueue.global().asyncAfter(deadline: .now() + duration + 1.0) {
            semaphore.signal()
        }
        semaphore.wait()
        
        stopPlayback()
    }
    
    // MARK: - Status Information
    
    func getPlaybackStatus() -> [String: Any] {
        bufferQueueLock.lock()
        let queueCount = bufferQueue.count
        bufferQueueLock.unlock()
        
        delayBufferLock.lock()
        let remainingDelay: TimeInterval
        if isDelayActive, let startTime = delayStartTime {
            remainingDelay = max(0, playbackDelay - Date().timeIntervalSince(startTime))
        } else {
            remainingDelay = 0
        }
        delayBufferLock.unlock()
        
        return [
            "isEngineRunning": isEngineRunning,
            "isPlayerPlaying": isPlayerPlaying,
            "volume": volume,
            "buffersScheduled": buffersScheduled,
            "buffersPlayed": buffersPlayed,
            "queuedBuffers": queueCount,
            "outputFormat": outputFormat?.description ?? "nil",
            "playbackDelay": playbackDelay,
            "isDelayActive": isDelayActive,
            "remainingDelay": remainingDelay
        ]
    }
}

// MARK: - AudioStreamDelegate Implementation
@available(macOS 13.0, *)
extension StreamingAudioPlayer: AudioStreamDelegate {
    
    public func audioStreamer(_ streamer: StreamingAudioRecorder, didReceive buffer: AVAudioPCMBuffer) {
        // Handle delay - don't even start the engine until delay expires
        if isDelayActive {
            delayBufferLock.lock()
            
            // Start delay timer on first buffer
            if delayStartTime == nil {
                delayStartTime = Date()
                print("\(timestamp()) StreamingAudioPlayer: Starting \(playbackDelay)s delay timer - audio engine will start after delay...")
            }
            
            // Check if delay has expired
            if isDelayExpired() {
                print("\(timestamp()) StreamingAudioPlayer: Delay period expired - starting audio engine now!")
                isDelayActive = false
                delayBufferLock.unlock()
                
                // Start playback fresh after delay
                if !isEngineRunning {
                    do {
                        try startPlayback()
                        print("\(timestamp()) StreamingAudioPlayer: Audio engine started fresh after delay")
                    } catch {
                        print("\(timestamp()) StreamingAudioPlayer: Failed to start playback: \(error)")
                        return
                    }
                }
                
                // Schedule this buffer immediately
                scheduleBuffer(buffer)
            } else {
                // Log delay progress
                let elapsed = Date().timeIntervalSince(delayStartTime!)
                let remaining = playbackDelay - elapsed
                
                if Int(elapsed * 10) % 5 == 0 { // Every 0.5 seconds
                    print("\(timestamp()) StreamingAudioPlayer: Delay in progress (\(String(format: "%.1f", remaining))s remaining) - engine not started")
                }
                
                delayBufferLock.unlock()
                return // Drop this buffer
            }
        } else {
            // No delay or delay has expired - ensure engine is running
            if !isEngineRunning {
                print("\(timestamp()) StreamingAudioPlayer: Starting playback on first audio buffer")
                do {
                    try startPlayback()
                } catch {
                    print("\(timestamp()) StreamingAudioPlayer: Failed to start playback: \(error)")
                    return
                }
            }
            
            // Schedule buffer for immediate playback
            scheduleBuffer(buffer)
        }
    }
    
    public func audioStreamer(_ streamer: StreamingAudioRecorder, didEncounterError error: Error) {
        print("\(timestamp()) StreamingAudioPlayer: Audio streamer error: \(error)")
    }
    
    public func audioStreamerDidFinish(_ streamer: StreamingAudioRecorder) {
        print("\(timestamp()) StreamingAudioPlayer: Audio streamer finished")
        stopPlayback()
    }
}